src/lxc/caps.c

```diff
@@ -126,6 +147,113 @@ int lxc_caps_up(void)
 	return 0;
 }
 
+int lxc_ambient_caps_up(void)
+{
+	int ret;
+	cap_t caps;
+	cap_value_t cap;
+	char *cap_names = NULL;
+
+	/* When we are run as root, we don't want to play with the capabilities. */
+	if (!getuid())
+		return 0;
+
+	caps = cap_get_proc();
+	if (!caps) {
+		SYSERROR("Failed to retrieve capabilities");
+		return -1;
+	}
+
+	for (cap = 0; cap <= CAP_LAST_CAP; cap++) {
+		cap_flag_value_t flag;
+
+		ret = cap_get_flag(caps, cap, CAP_PERMITTED, &flag);
+		if (ret < 0) {
+			if (errno == EINVAL) {
+				INFO("Last supported cap was %d", cap - 1);
+				break;
+			}
+
+			SYSERROR("Failed to retrieve capability flag");
+			goto out;
+		}
+
+		ret = cap_set_flag(caps, CAP_INHERITABLE, 1, &cap, flag);
+		if (ret < 0) {
+			SYSERROR("Failed to set capability flag");
+			goto out;
+		}
+	}
+
+	ret = cap_set_proc(caps);
+	if (ret < 0) {
+		SYSERROR("Failed to set capabilities");
+		goto out;
+	}
+
+	for (cap = 0; cap <= CAP_LAST_CAP; cap++) {
+		ret = prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_RAISE, cap, 0, 0);
+		if (ret < 0) {
+			WARN("%s - Failed to raise ambient capability %d",
+			     strerror(errno), cap);
+			goto out;
+		}
+	}
+
+	cap_names = cap_to_text(caps, NULL);
+	if (!cap_names)
+		goto out;
+
+	TRACE("Raised %s in inheritable and ambient capability set", cap_names);
+
+out:
+
+	cap_free(cap_names);
```

Comment:
shouldn't this just be free(cap_names) ?

Reply:
According to the manpage you're supposed to use `cap_free()`:
```
 cap_to_text()  converts the capability state in working storage identified
       by cap_p into a nul-terminated human-readable string.  This function allo‐
       cates any memory necessary to contain the string, and returns a pointer to
       the string.  If the pointer len_p is not NULL,  the  function  shall  also
       return the full length of the string (not including the nul terminator) in
       the location pointed to by len_p.  The capability state in  working  stor‐
       age,  identified  by  cap_p,  is  completely  represented in the character
       string.  When the  capability  state  in  working  storage  is  no  longer
       required,  the  caller  should  free  any  releasable  memory  by  calling
       cap_free() with the returned string pointer as an argument.
```

Reply:
Hah - good call, thanks.